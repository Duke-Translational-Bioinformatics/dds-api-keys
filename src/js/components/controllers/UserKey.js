import { connect } from 'react-redux'
import { setUserApiKey } from '../../actions'
import ManageKey from '../views/ManageKey'

const mapStateToProps = state => ({
  userApiKey: state.userApiKey
})

const mapDispatchToProps = dispatch => ({
  setUserApiKey: userApiKey => dispatch(setUserApiKey(userApiKey)),
  destroyUserApiKey: () => dispatch(setUserApiKey(null))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ManageKey)
