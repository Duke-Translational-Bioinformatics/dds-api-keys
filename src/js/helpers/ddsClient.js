import axios from "axios";
import config from "../config/ddsconfig.js";

const ddsApiBaseURL = () => config["dds_api_base_url"];

var ddsClient = {
  send(payload, processData, handleFailure) {
    axios(payload)
      .then(processData)
      .catch((error) => {
        handleFailure(error.response.data);
      })
  },

  getJwtToken(accessToken, handleToken, handleFailure) {
    var url = `${ddsApiBaseURL}/user/api_token`;
    this.send(
      {
        url: url,
        method: 'get',
        params: {access_token: accessToken}
      },
      (response) => {
        var thisData = response.data;
        handleToken(thisData.api_token, thisData.expires_on, thisData.time_to_live);
      },
      handleFailure
    )
  },

  getCurrentUser(jwtToken, handleUser, handleFailure) {
    var url = `${ddsApiBaseURL}/current_user`
    this.send(
      {
        url: url,
        method: 'get',
        headers: {Authorization: jwtToken}
      },
      (response) => { handleUser(response.data) },
      handleFailure
    )
  },

  getUserApiKey(jwtToken, handleUserApiKey, handleFailure) {
    var url = `${ddsApiBaseURL}/current_user/api_key`
    this.send(
      {
        url: url,
        method: 'get',
        headers: {Authorization: jwtToken}
      },
      (response) => { handleUserApiKey(response.data.key) },
      handleFailure
    )
  },

  setUserApiKey(jwtToken, handleUserApiKey, handleFailure) {
    var url = `${ddsApiBaseURL}/current_user/api_key`
    this.send(
      {
        url: url,
        method: 'put',
        headers: {Authorization: jwtToken}
      },
      (response) => { handleUserApiKey(response.data.key) },
      handleFailure
    )
  },

  destroyUserApiKey(jwtToken, handleFailure) {
    var url = `${ddsApiBaseURL}/current_user/api_key`
    this.send(
      {
        url: url,
        method: 'delete',
        headers: {Authorization: jwtToken}
      },
      () => {},
      handleFailure
    )
  }
};
export default ddsClient;
