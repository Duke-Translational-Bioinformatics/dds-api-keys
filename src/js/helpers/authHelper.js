import Axios from "axios";
import ddsClient from './ddsClient'
import config from "../config/authconfig.js";

const oauthRedirect = config["oauth_redirect"];
const tokenStoreKey = "api-keys-token";
const tokenExpirationStoreKey = "api-keys-token-expiration"

const authHelper = {
  isLoggedIn() {
    if(this.jwt() && this.jwtExpiration()) {
      if(parseInt(this.jwtExpiration()) >= Date.now()) {
        return true;
      }
      else {
        sessionStorage.clear();
        return false;
      }
    }
  },

  jwt() {
    return sessionStorage.getItem(tokenStoreKey);
  },

  jwtExpiration() {
    return sessionStorage.getItem(tokenExpirationStoreKey);
  },

  accessTokenExists() {
    this.accessToken = this.getOauthCodeFromURI();
    return this.accessToken != null;
  },

  getOauthCodeFromURI() {
    if (window.location.href.indexOf("#access_token") > 0) {
      const parts = window.location.hash.slice(1).split("&");
      const oauthCode = parts[0].split("=")[1];
      return oauthCode;
    }
  },

  getOauthProviderInfo(successHandler, errorHandler) {
    ddsClient.getDefaultOauthProvider(
      (provider) => {
        if(provider) {
          successHandler(
            `${provider.login_initiation_url}&state=login&redirect_uri=${
              oauthRedirect
            }`
          );
        }
        else {
          errorHandler(
            "No Default StorageProvider returned from DDS!"
          );
        }
      },
    (errorResponse) => {
      errorHandler(
        "Error getting default auth provider information " + errorResponse.message
      );
    });
  },

  login() {
    return new Promise((resolve, reject) => {
      if (this.isLoggedIn()) {
        resolve(true);
      }
      else {
        if(this.accessTokenExists()) {
          ddsClient.getJwtToken(
            this.accessToken,
            (jwtToken, expiration, timeToLive) => {
              sessionStorage.setItem(tokenStoreKey, jwtToken);
              sessionStorage.setItem(tokenExpirationStoreKey, Date.now() + timeToLive);
              window.location.replace("#");
              if (typeof window.history.replaceState == 'function') {
                history.replaceState({}, '', window.location.href.slice(0, -1));
              }
              resolve(true);
            },
            (errorMessage) => {
              reject("Could not get JwtToken: " + errorMessage);
            }
          );
        }
        else {
          this.getOauthProviderInfo(
            (dukeOauthUrl) => {
              window.location.assign(dukeOauthUrl);
              resolve(true);
            },
            (errorMessage) => {
              reject(errorMessage);
            }
          );
        }
      }
    });
  }
};

export default authHelper;
