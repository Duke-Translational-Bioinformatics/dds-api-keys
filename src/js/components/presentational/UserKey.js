import React, { Component } from "react";
import PropTypes from 'prop-types'
import ApiKey from "../container/ApiKey"
import authHelper from '../../helpers/authHelper';
import ddsClient from '../../helpers/ddsClient';

class UserKey extends Component {
  constructor(props) {
    super(props);
    this.handleAuthenticationSuccess = this.handleAuthenticationSuccess.bind(this);
    this.handleCurrentUser = this.handleCurrentUser.bind(this);
  }

  componentWillMount() {
    if (!authHelper.isLoggedIn()) {
      authHelper.login().then(
        this.handleAuthenticationSuccess,
        this.handleException
      );
    }
  }

  handleAuthenticationSuccess(isSuccessful) {
    var jwtToken = authHelper.jwt();
    ddsClient.getCurrentUser(
      jwtToken,
      this.handleCurrentUser,
      this.handleException
    )
  }

  handleCurrentUser(user) {
    this.props.setCurrentUser(user);
  }

  render() {
    if (authHelper.isLoggedIn()) {
      return (
        <div>
          <p>User { this.props.currentUser.full_name }</p>
          <ApiKey />
        </div>
      )
    }
    else {
      return (
        <div>
          <p>Fetching User</p>
        </div>
      )
    }
  }
}

UserKey.propTypes = {
  currentUser: PropTypes.object,
  setCurrentUser: PropTypes.func.isRequired
}
export default UserKey;
