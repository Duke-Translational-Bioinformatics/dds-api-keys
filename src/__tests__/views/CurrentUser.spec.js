jest.useFakeTimers();
import React from 'react';
import { shallow } from 'enzyme';
import CurrentUser from 'js/views/CurrentUser';
import UserKey from "js/controllers/UserKey"
import { ThemeProvider } from "styled-components";
import { Modal, theme } from "dracs";

import authHelper from 'js/helpers/authHelper';
import ddsClient from 'js/helpers/ddsClient';

describe('CurrentUser View', () => {
  let wrapper;
  let mockSetCurrentUser = jest.fn();

  const origIsLoggedInF = authHelper.isLoggedIn;
  let userIsLoggedIn = false;

  const expectedCurrentUser = {
    full_name: 'user name'
  };

  function mockIsLoggedIn() {
    return userIsLoggedIn;
  }

  const origLoginF = authHelper.login;
  const origJwtF = authHelper.jwt;
  const fakeJwt = 'abc123xyz';
  const origGetCurrentUserF = ddsClient.getCurrentUser;

  beforeEach(() => {
    mockSetCurrentUser.mockClear();
  });

  describe('UI', () => {
    var tearDownUIMocks;

    afterEach(() => {
      tearDownUIMocks();
    });

    describe('when user is already logged in', () => {
      let setUpUIMocks = () => {
        authHelper.isLoggedIn = jest.fn();
        authHelper.isLoggedIn.mockImplementation(mockIsLoggedIn);

        authHelper.login = jest.fn();
      }

      tearDownUIMocks = () => {
        authHelper.isLoggedIn = origIsLoggedInF
        authHelper.login = origLoginF;
        userIsLoggedIn = false;
      }

      it('should render information about the user and a UserKey controller', () => {
        userIsLoggedIn = true;
        setUpUIMocks();

        expect(authHelper.isLoggedIn()).toBeTruthy();
        wrapper = shallow(<CurrentUser currentUser={expectedCurrentUser} setCurrentUser={mockSetCurrentUser} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(Modal)).toHaveProp('active', false);
        expect(wrapper.state().hasError).toBeFalsy();
      });
    });


    describe('when user is not already logged in', () => {
      let mockedPromiseResolver = {
        then: jest.fn()
      };
      function mockedPromise() {
        return mockedPromiseResolver
      }

      let setUpUIMocks = () => {
        authHelper.isLoggedIn = jest.fn();
        authHelper.isLoggedIn.mockImplementation(mockIsLoggedIn);

        authHelper.login = jest.fn();
        authHelper.login.mockImplementation(mockedPromise);
      }

      tearDownUIMocks = () => {
        authHelper.isLoggedIn = origIsLoggedInF
        authHelper.login = origLoginF;
        userIsLoggedIn = false;
      }

      describe('when a login error occurs', () => {
        it('should render an initialization as it attempts to log the user in', () => {
          let thisMessage = {error: "404", message: "got an error"};
          setUpUIMocks();
          expect(authHelper.isLoggedIn()).toBeFalsy();

          wrapper = shallow(
            <CurrentUser setCurrentUser={mockSetCurrentUser} />
          );
          let wrapperWithError = wrapper.setState({
            hasError: true,
            errorMessage: thisMessage
          });
          expect(wrapperWithError).toMatchSnapshot();
          expect(wrapperWithError.find(Modal)).toHaveProp('active', true);
          expect(wrapperWithError.state().hasError).toBeTruthy();
        });
      });

      describe('when a a login error is not encountered', () => {
        it('should render an initialization as it attempts to log the user in', done => {
          setUpUIMocks();
          expect(authHelper.isLoggedIn()).toBeFalsy();

          wrapper = mount(
            <ThemeProvider theme={theme}>
              <CurrentUser setCurrentUser={mockSetCurrentUser} />
            </ThemeProvider>
          );
          expect(wrapper).toMatchSnapshot();
          const subject = wrapper.find("CurrentUser");
          setImmediate(() => {
            expect(authHelper.login).toBeCalled();
            expect(mockedPromiseResolver.then).toHaveBeenCalledWith(
              subject.instance().handleAuthenticationSuccess,
              subject.instance().handleException
            );
            done();
          });
        });
      });
    });
  });

  describe('Handler Functions', () => {
    let props = {
      setCurrentUser: mockSetCurrentUser
    };
    let subject = new CurrentUser(props);

    describe('handleException', () => {
      describe('when refs.current_user_rendered is present', () => {
        it('should set the hasError state', () => {
          window.alert = jest.fn();
          let thisMessage = {error: "404", message: "got an error"};
          wrapper = mount(
            <ThemeProvider theme={theme}>
              <CurrentUser currentUser={expectedCurrentUser} setCurrentUser={mockSetCurrentUser} />
            </ThemeProvider>
          );
          const renderedCurrentUser = wrapper.find("CurrentUser");
          expect(renderedCurrentUser.state()).toEqual({
            hasError: false
          });
          expect(renderedCurrentUser.instance().refs.current_user_rendered).toBeTruthy();
          renderedCurrentUser.instance().handleException(thisMessage);
          expect(renderedCurrentUser.state()).toEqual({
            hasError: true,
            errorMessage: thisMessage
          });
        });
      });

      describe('when refs.current_user_rendered is absent', () => {
        it('should not set the hasError state', () => {
          let thisMessage = {error: "404", message: "got an error"};
          expect(subject.state).toEqual({
            hasError: false
          });
          expect(subject.refs.current_user_rendered).toBeFalsy();
          subject.handleException(thisMessage);
          expect(subject.state).toEqual({
            hasError: false
          });
        });
      });
    });

    describe('acknowlegeException', () => {
      describe('when refs.current_user_rendered is present', () => {
        it('should clear any errors present', () => {
          window.alert = jest.fn();
          let thisMessage = {error: "404", message: "got an error"};
          wrapper = mount(
            <ThemeProvider theme={theme}>
              <CurrentUser currentUser={expectedCurrentUser} setCurrentUser={mockSetCurrentUser} />
            </ThemeProvider>
          );
          const renderedCurrentUser = wrapper.find("CurrentUser");
          expect(renderedCurrentUser.state()).toEqual({
            hasError: false
          });
          expect(renderedCurrentUser.instance().refs.current_user_rendered).toBeTruthy();
          renderedCurrentUser.instance().handleException(thisMessage);
          expect(renderedCurrentUser.state()).toEqual({
            hasError: true,
            errorMessage: thisMessage
          });

          renderedCurrentUser.instance().acknowlegeException();
          expect(renderedCurrentUser.state()).toEqual({
            hasError: false
          });
        });
      });

      describe('when refs.current_user_rendered is absent', () => {
        it('should not reset state', () => {
          const origSetStateF = subject.setState;
          subject.setState = jest.fn();
          expect(subject.refs.current_user_rendered).toBeFalsy();
          subject.acknowlegeException();
          expect(subject.setState).not.toHaveBeenCalled();
          subject.setState = origSetStateF;
        });
      });
    });

    describe('handleAuthenticationSuccess', () => {
      it('should get the jwt and attempt to get the currentUser', () => {
        authHelper.jwt = jest.fn();
        authHelper.jwt.mockImplementation(() => {
          return fakeJwt;
        });
        ddsClient.getCurrentUser = jest.fn();
        subject.handleAuthenticationSuccess(true);
        expect(ddsClient.getCurrentUser).toHaveBeenCalledWith(
          fakeJwt,
          subject.handleCurrentUser,
          subject.ignorePrematureCallException
        )
        ddsClient.getCurrentUser = origGetCurrentUserF;
        authHelper.jwt = origJwtF;
      });
    });

    describe('ignorePrematureCallException', () => {
      describe('when authHelper.jwt is defined', () => {
        it('should forward the exception to handleException', () => {
          let thisMessage = {error: "404", message: "got an error"};
          authHelper.jwt = jest.fn();
          authHelper.jwt.mockImplementation(() => {
            return fakeJwt;
          });
          let origHandleExceptionF = subject.handleException;
          subject.handleException = jest.fn();
          expect(authHelper.jwt()).toBeTruthy();
          subject.ignorePrematureCallException(thisMessage);
          expect(subject.handleException).toHaveBeenCalledWith(thisMessage);
          subject.handleException = origHandleExceptionF;
          authHelper.jwt = origJwtF;
        });
      });

      describe('when authHelper.jwt is not defined', () => {
        it('should not forward the exception to handleException', () => {
          let thisMessage = {error: "404", message: "got an error"};
          let origHandleExceptionF = subject.handleException;
          subject.handleException = jest.fn();
          expect(authHelper.jwt()).toBeFalsy();
          subject.ignorePrematureCallException(thisMessage);
          expect(subject.handleException).not.toHaveBeenCalled();
          subject.handleException = origHandleExceptionF;
        });
      });
    });

    describe('handleCurrentUser', () => {
      it('should call the setCurrentUser prop', () => {
        subject.handleCurrentUser(expectedCurrentUser);
        expect(mockSetCurrentUser).toHaveBeenCalledWith(expectedCurrentUser);
      });
    });
  });
});
