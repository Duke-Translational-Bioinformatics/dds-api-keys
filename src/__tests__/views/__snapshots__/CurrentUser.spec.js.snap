// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentUser View UI when user is already logged in should render information about the user, an problem alert Modal, and a UserKey controller 1`] = `
<div>
  <ActivableRenderer
    active={false}
    delay={300}
    id="login_api_problem_notification"
    onEscKeyDown={[Function]}
    type="medium"
  >
    <typography__H4>
      A Problem has occurred
    </typography__H4>
    <typography__P />
    <Button
      label="OK"
      onClick={[Function]}
    />
  </ActivableRenderer>
  <Card
    border={true}
    height="400px"
    raised={true}
    width="500px"
  >
    <CardHeader
      border={true}
      icon={
        <div
          className="item-row"
        >
          <typography__H4
            bold={true}
            color="#FFFFFF"
          >
            Duke Data Service User Secret
          </typography__H4>
          <typography__P
            className="header-title"
            color="#FFFFFF"
          >
            user name
          </typography__P>
        </div>
      }
      style={
        Object {
          "backgroundColor": "#0680CD",
          "maxWidth": "500px",
        }
      }
    />
    <CardBody>
      <Connect(ManageKey) />
    </CardBody>
  </Card>
</div>
`;

exports[`CurrentUser View UI when user is not already logged in when a a login error is not encountered should render an initialization as it attempts to log the user in 1`] = `
<ThemeProvider
  theme={
    Object {
      "boxShadow": Object {
        "button": "0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)",
        "card": "0 1px 3px 0 rgba(0, 0, 0, 0.25)",
        "header": "0 0 12px 0 rgba(0, 0, 0, 0.40)",
        "shadow2px": "0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 3px 1px -2px rgba(0, 0, 0, 0.2), 0 1px 5px 0 rgba(0, 0, 0, 0.12)",
        "shadow3px": "0 3px 4px 0 rgba(0, 0, 0, 0.14), 0 3px 3px -2px rgba(0, 0, 0, 0.2), 0 1px 8px 0 rgba(0, 0, 0, 0.12)",
        "zdepth5": "0 19px 60px rgba(0, 0, 0, 0.3), 0 15px 20px rgba(0, 0, 0, 0.22)",
      },
      "colors": Object {
        "action": "#0680cd",
        "action87pct": "hsla(203.200000000005, 94.3%, 41.4%, 0.87)",
        "actionHover": "#53a4dc",
        "base": "#131516",
        "base45pct": "hsla(0,0%,14.9%,0.44999999999999996)",
        "base57pct": "hsla(0, 0%, 14.9%, 0.5700000000000001)",
        "base90pct": "hsla(0,0%,14.9%,0.9)",
        "bg": "#fff",
        "border": "#e8e5e2",
        "borderDark": "#d3d3d3",
        "darkBlue": "#036",
        "dashBG": "#f6f5f4",
        "error": "#cc3300",
        "error87pct": "hsla(15,100%,40%,0.87)",
        "muted": "#696969",
        "statusGreen": "#008000",
        "subheading": "#545454",
        "successGreen": "#4AC948",
        "tutorialHighlight": "rgb(255,215,0)",
        "warning": "#ffd960",
      },
      "flexboxgrid": Object {
        "breakpoints": Object {
          "lg": 90,
          "md": 73.125,
          "sm": 48,
          "xs": 0,
        },
        "container": Object {
          "lg": 76,
          "md": 61,
          "sm": 46,
        },
        "gutterWidth": 1,
        "outerMargin": 1,
      },
      "sizing": Object {
        "base": 8,
        "borderRadius": 4,
        "fontSize": 16,
      },
      "zIndex": Object {
        "lg": 500,
        "md": 100,
        "s": 10,
        "xl": 1000,
        "xs": 1,
      },
    }
  }
>
  <CurrentUser
    setCurrentUser={[MockFunction]}
  >
    <div>
      <ActivableRenderer
        active={false}
        delay={300}
        id="login_api_problem_notification"
        onEscKeyDown={[Function]}
        type="medium"
      />
      <Card
        border={true}
        height="400px"
        raised={true}
        width="500px"
      >
        <card__CardWrapper
          border={true}
          height="400px"
          raised={true}
          width="500px"
        >
          <section
            className="card__CardWrapper-aKCya fwnmPO"
            height="400px"
            width="500px"
          >
            <CardHeader
              border={true}
              className=""
              dragHandle={false}
              htmlTitle={false}
              icon={
                <typography__H4
                  bold={true}
                  color="#FFFFFF"
                >
                  Duke Data Service User Secret
                </typography__H4>
              }
              style={
                Object {
                  "backgroundColor": "#0680CD",
                  "maxWidth": "500px",
                }
              }
              title=""
            >
              <header__CardHeaderWrap
                border={true}
                className=""
                dragHandle={false}
                style={
                  Object {
                    "backgroundColor": "#0680CD",
                    "maxWidth": "500px",
                  }
                }
              >
                <header
                  className="header__CardHeaderWrap-gWktLy joQkYb"
                  style={
                    Object {
                      "backgroundColor": "#0680CD",
                      "maxWidth": "500px",
                    }
                  }
                >
                  <typography__H4
                    bold={true}
                    color="#FFFFFF"
                  >
                    <h4
                      className="typography__H4-ggBxhx clfzTd"
                      color="#FFFFFF"
                    >
                      Duke Data Service User Secret
                    </h4>
                  </typography__H4>
                </header>
              </header__CardHeaderWrap>
            </CardHeader>
            <CardBody
              className=""
              id=""
              padding=""
              style={Object {}}
            >
              <body__CardBodyWrap
                className=""
                id=""
                padding=""
                style={Object {}}
              >
                <article
                  className="body__CardBodyWrap-gxNncO iKghek"
                  id=""
                  style={Object {}}
                >
                  <typography__P>
                    <p
                      className="typography__P-jyzngc gYZjzX"
                    >
                      Initializing ... 
                    </p>
                  </typography__P>
                  <Spinner
                    color="#0680cd"
                    margin="4px"
                    size={32}
                  >
                    <spinner__ElementWrap
                      margin="4px"
                      size={32}
                    >
                      <div
                        className="spinner__ElementWrap-hXnaGg eIYEol"
                        size={32}
                      >
                        <spinner__SpinnerWrap
                          size={32}
                        >
                          <div
                            className="spinner__SpinnerWrap-cJPLcw eSvGES"
                            size={32}
                          >
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="0"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-1.2s",
                                  "WebkitTransform": "rotate(0deg) translate(146%)",
                                  "animationDelay": "-1.2s",
                                  "transform": "rotate(0deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-1.2s",
                                    "WebkitTransform": "rotate(0deg) translate(146%)",
                                    "animationDelay": "-1.2s",
                                    "transform": "rotate(0deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="1"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-1.1s",
                                  "WebkitTransform": "rotate(30deg) translate(146%)",
                                  "animationDelay": "-1.1s",
                                  "transform": "rotate(30deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-1.1s",
                                    "WebkitTransform": "rotate(30deg) translate(146%)",
                                    "animationDelay": "-1.1s",
                                    "transform": "rotate(30deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="2"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-1s",
                                  "WebkitTransform": "rotate(60deg) translate(146%)",
                                  "animationDelay": "-1s",
                                  "transform": "rotate(60deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-1s",
                                    "WebkitTransform": "rotate(60deg) translate(146%)",
                                    "animationDelay": "-1s",
                                    "transform": "rotate(60deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="3"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.9s",
                                  "WebkitTransform": "rotate(90deg) translate(146%)",
                                  "animationDelay": "-0.9s",
                                  "transform": "rotate(90deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.9s",
                                    "WebkitTransform": "rotate(90deg) translate(146%)",
                                    "animationDelay": "-0.9s",
                                    "transform": "rotate(90deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="4"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.8s",
                                  "WebkitTransform": "rotate(120deg) translate(146%)",
                                  "animationDelay": "-0.8s",
                                  "transform": "rotate(120deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.8s",
                                    "WebkitTransform": "rotate(120deg) translate(146%)",
                                    "animationDelay": "-0.8s",
                                    "transform": "rotate(120deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="5"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.7s",
                                  "WebkitTransform": "rotate(150deg) translate(146%)",
                                  "animationDelay": "-0.7s",
                                  "transform": "rotate(150deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.7s",
                                    "WebkitTransform": "rotate(150deg) translate(146%)",
                                    "animationDelay": "-0.7s",
                                    "transform": "rotate(150deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="6"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.6s",
                                  "WebkitTransform": "rotate(180deg) translate(146%)",
                                  "animationDelay": "-0.6s",
                                  "transform": "rotate(180deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.6s",
                                    "WebkitTransform": "rotate(180deg) translate(146%)",
                                    "animationDelay": "-0.6s",
                                    "transform": "rotate(180deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="7"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.5s",
                                  "WebkitTransform": "rotate(210deg) translate(146%)",
                                  "animationDelay": "-0.5s",
                                  "transform": "rotate(210deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.5s",
                                    "WebkitTransform": "rotate(210deg) translate(146%)",
                                    "animationDelay": "-0.5s",
                                    "transform": "rotate(210deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="8"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.4s",
                                  "WebkitTransform": "rotate(240deg) translate(146%)",
                                  "animationDelay": "-0.4s",
                                  "transform": "rotate(240deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.4s",
                                    "WebkitTransform": "rotate(240deg) translate(146%)",
                                    "animationDelay": "-0.4s",
                                    "transform": "rotate(240deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="9"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.3s",
                                  "WebkitTransform": "rotate(270deg) translate(146%)",
                                  "animationDelay": "-0.3s",
                                  "transform": "rotate(270deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.3s",
                                    "WebkitTransform": "rotate(270deg) translate(146%)",
                                    "animationDelay": "-0.3s",
                                    "transform": "rotate(270deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="10"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.2s",
                                  "WebkitTransform": "rotate(300deg) translate(146%)",
                                  "animationDelay": "-0.2s",
                                  "transform": "rotate(300deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.2s",
                                    "WebkitTransform": "rotate(300deg) translate(146%)",
                                    "animationDelay": "-0.2s",
                                    "transform": "rotate(300deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                            <spinner__SpinnerBar
                              color="#0680cd"
                              key="11"
                              style={
                                Object {
                                  "WebkitAnimationDelay": "-0.1s",
                                  "WebkitTransform": "rotate(330deg) translate(146%)",
                                  "animationDelay": "-0.1s",
                                  "transform": "rotate(330deg) translate(146%)",
                                }
                              }
                            >
                              <div
                                className="spinner__SpinnerBar-gVLYMq dbGfAr"
                                color="#0680cd"
                                style={
                                  Object {
                                    "WebkitAnimationDelay": "-0.1s",
                                    "WebkitTransform": "rotate(330deg) translate(146%)",
                                    "animationDelay": "-0.1s",
                                    "transform": "rotate(330deg) translate(146%)",
                                  }
                                }
                              />
                            </spinner__SpinnerBar>
                          </div>
                        </spinner__SpinnerWrap>
                      </div>
                    </spinner__ElementWrap>
                  </Spinner>
                </article>
              </body__CardBodyWrap>
            </CardBody>
          </section>
        </card__CardWrapper>
      </Card>
    </div>
  </CurrentUser>
</ThemeProvider>
`;

exports[`CurrentUser View UI when user is not already logged in when a login error occurs should render an initialization as it attempts to log the user in 1`] = `
<div>
  <ActivableRenderer
    active={true}
    delay={300}
    id="login_api_problem_notification"
    onEscKeyDown={[Function]}
    type="medium"
  >
    <typography__H4>
      A Problem has occurred
    </typography__H4>
    <typography__P>
      {"error":"404","message":"got an error"}
    </typography__P>
    <Button
      label="OK"
      onClick={[Function]}
    />
  </ActivableRenderer>
  <Card
    border={true}
    height="400px"
    raised={true}
    width="500px"
  >
    <CardHeader
      border={true}
      className=""
      dragHandle={false}
      htmlTitle={false}
      icon={
        <typography__H4
          bold={true}
          color="#FFFFFF"
        >
          Duke Data Service User Secret
        </typography__H4>
      }
      style={
        Object {
          "backgroundColor": "#0680CD",
          "maxWidth": "500px",
        }
      }
      title=""
    />
    <CardBody
      className=""
      id=""
      padding=""
      style={Object {}}
    >
      <typography__P>
        Initializing ... 
      </typography__P>
      <Spinner
        color="#0680cd"
        margin="4px"
        size={32}
      />
    </CardBody>
  </Card>
</div>
`;
